lisa/analysis/rta.py

```diff
@@ -0,0 +1,400 @@
+          * An ``event``  column: the generated event
+          * A  ``phase``  column: the phases counter for each ``__pid``:``__comm`` task
+          * A  ``phase_loop``  colum: the phase_loops's counter
+          * A  ``thread_loop`` column: the thread_loop's counter
+
+        The ``event`` column can report these events:
+
+          * ``start``: the start of the ``__pid``:``__comm`` related event
+          * ``end``: the end of the ``__pid``:``__comm`` related event
+
+        """
+        df = self.trace.df_events('rtapp_loop')
+        return self._task_filtered(df, task)
+
+    @requires_events('rtapp_task')
+    def tasks_window(self, task):
+        """
+        Return the start end end time for the specified task.
+
+        :param task: the (optional) rt-app task to filter for
+        :type task: int or str or :class:`TaskID`
+        """
+        task_df = self.df_rtapp_task(task)
+        start_time = task_df[task_df.event == "start"].index[0]
+        end_time = task_df[task_df.event == "end"].index[0]
+
+        return (start_time, end_time)
+
+    ###########################################################################
+    # rtapp_phase events related methods
+    ###########################################################################
+
+    @requires_events('rtapp_event')
+    def df_rtapp_event(self, task=None):
+        """
+        Returns a :class:`pandas.DataFrame` of all the rt-app generated events.
+
+        :param task: the (optional) rt-app task to filter for
+        :type task: int or str or :class:`TaskID`
+
+        :returns: a :class:`pandas.DataFrame` with:
+
+          * A  ``__comm`` column: the actual rt-app trace task name
+          * A  ``__pid``  column: the PID of the task
+          * A ``__cpu``  column: the CPU on which the task was running at event
+                                 generation time
+          * A ``__line`` column: the ftrace line numer
+          * A ``type`` column: the type of the generated event
+          * A ``desc`` column: the mnemonic type of the generated event
+          * A ``id`` column: the ID of the resource associated to the event,
+                             e.g. the ID of the fired timer
+
+        The ``index`` represents the timestamp of the event.
+        """
+        df = self.trace.df_events('rtapp_event')
+        return self._task_filtered(df, task)
+
+    ###########################################################################
+    # rtapp_stats events related methods
+    ###########################################################################
+
+    @memoized
+    @requires_events('rtapp_stats')
+    def _get_stats(self):
+        df = self.trace.df_events('rtapp_stats')
+        # Add performance metrics column, performance is defined as:
+        #             slack
+        #   perf = -------------
+        #          period - run
+        df.loc[:,'perf_index'] = df['slack'] / (df['c_period'] - df['c_run'])
```

Comment:
From a quick experiment, `df.loc[:,'perf_index'] = ...` seems to do the same thing as `df[perf_index'] = ...`, which is the regular way of adding a column.

Reply:
No, that's not the same and if you run on a notebook you get a bold warning that you should better use the `.loc` notation.

Reply:
interesting, do you have the warning in question ? We will have to propagate that update to the whole LISA if that's really the case ...

Reply:
Something like:
```
2019-09-17 16:57:27,277 WARNING : py.warnings  : /data/Code/lisa/.lisa-venv-3.5/lib/python3.5/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
```

Reply:
Maybe that happens only if we add columns to a slice/view... not to the entire original DF?
But in that case, still, when we use `trace.df_events()` and trace is a view we will have the same issue, don't we?

Reply:
AFAIK the solution they provide is meant to fix code that intends to modify the dataframe directly:
```
# What the warning assumes
df[row_indexer][col_indexer] = ...
do_something(df[col_indexer]) # that column does not exists, since it was set on df[row_indexer] and not df.
# so the warning tells how to fix that case

# what we have
df = ... # in another function
df_or_slice = either df or df[row_indexer] # in another function

df_or_slice[col_indexer] = ...
do_something(df_or_slice[col_indexer])
# we never want to modify df itself, since what we manage at this level is df_or_slice
```

Reply:
Do you have the issue even with the `copy()` of the df returned by `df_events()` in the current version of the code ? `copy()` returns a Dataframe, so it would surprising to get a warning about a `copy of a slice`.

Reply:
>  But in that case, still, when we use trace.df_events() and trace is a view we will have the same issue, don't we?

Yes, `trace.df_events()` can return either a slice or a dataframe, which is not an issue since we consider this data readonly and we never attempt to modify it ;)

Reply:
So, you mean we should go with a deep copy? Because in that case, we are back to the point where we will have a memory overhead.

Reply:
Ok. let's got with a deep copy to be on the safe side and keep read-only the "raw" dataframe.